### PR TITLE
WIP backup/recovery - not fully tested

### PR DESCRIPTION
**Pull request by asalkeld**
I have only tested bits:
- the backup download container is working nicely to retrieve the snapshot
- I am using quay.io/asalkeld/getbackup:latest (I dont' have permission to quay.io/openshift-azure/)
- next step is to test etcdctl snapshot restore
- then test end-to-end

Note: 
- the backup is not working as I would expect. It *is* running every 10 minutes but it keeps writing over blob "etcd/backup" (you can see the timestamp getting updated)

/hold

### CODE REVIEW

File: Dockerfile.getbackup [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FROM centos:7
+RUN yum -y update
+COPY getbackup .
+ENTRYPOINT [ "/getbackup" ]
+# example:
+# docker run -it -v /etc/origin/cloudprovider/:/_data/_out -v /var/lib/etcd:/out:z quay.io/asalkeld/getbackup:dbbeeaaa --blobname=etcd/backup --destination=/out/backup
```
Comment (kargakis) [inline Dockerfile.getbackup]: Add this somewhere in the readme instead of here

Comment (kargakis) [reply Dockerfile.getbackup]: Or remove altogether

File: cmd/getbackup/getbackup.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/openshift/openshift-azure/pkg/log"
+	"github.com/openshift/openshift-azure/pkg/util/azureclient"
+	azureclientstorage "github.com/openshift/openshift-azure/pkg/util/azureclient/storage"
+	"github.com/openshift/openshift-azure/pkg/util/nodeconf"
+)
+
+var (
+	logLevel    = flag.String("loglevel", "Debug", "valid values are Debug, Info, Warning, Error")
```
Comment (kargakis) [inline cmd/getbackup/getbackup.go]: Keep the help messages across all flags consistent: start with a capital letter.

File: cmd/getbackup/getbackup.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	interval    = flag.Duration("interval", 5*time.Second, "How often to retry the download.")
+	blobName    = flag.String("blobname", "", "name of the blog to download")
```
Comment (kargakis) [inline cmd/getbackup/getbackup.go]: s/blog/blob/

File: pkg/arm/arm.go [hunk 4]
```diff
@@ -54,12 +54,15 @@ func (g *simpleGenerator) Generate(ctx context.Context, cs *api.OpenShiftManaged
 	if err != nil {
 		return nil, err
 	}
+	log.Infof("*** IsRecovery: %v, BackupBlobName: %v", len(g.pluginConfig.RecoverEtcdFromBackup) > 0, g.pluginConfig.RecoverEtcdFromBackup)
```
Comment (kargakis) [inline pkg/arm/arm.go]: remove

File: pkg/fakerp/fakerp.go [hunk 5]
```diff
@@ -94,7 +92,7 @@ func CreateOrUpdate(ctx context.Context, oc *v20180930preview.OpenShiftManagedCl
 
 	// write out development files
 	log.Info("write helpers")
-	err = writeHelpers(cs, azuretemplate)
```
Comment (kargakis) [inline pkg/fakerp/fakerp.go]: revert this change since it's only used in here.

File: pkg/util/nodeconf/nodeconf.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	b, err = yaml.Marshal(c.Config.AdminKubeconfig)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile("_data/_out/admin.kubeconfig", b, 0600)
+}
+
+// GetAzureConf returns the azure credentials to be used by sync and getbackup
+func GetAzureConf() (map[string]string, error) {
+	b, err := ioutil.ReadFile("_data/_out/azure.conf")
```
Comment (kargakis) [inline pkg/util/nodeconf/nodeconf.go]: If we want this in a shared library, I'd probably make _data configurable.

File: cmd/getbackup/getbackup.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+		return err
+	}
+	defer rc.Close()
+	logrus.Print("read blob")
+
+	b, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return err
+	}
+	logrus.Print("creating ", destPath)
```
Comment (kargakis) [inline cmd/getbackup/getbackup.go]: prefer using string formatting, also is consistent with what you do elsewhere

File: cmd/getbackup/getbackup.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+	df, err := os.Create(destPath)
+	if err != nil {
+		return err
+	}
+	defer df.Close()
+	_, err = df.Write(b)
+	logrus.Print("Write returned ", err)
```
Comment (kargakis) [inline cmd/getbackup/getbackup.go]: unnecessary, the error is handled already in the caller

File: cmd/getbackup/getbackup.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	return err
+}
+
+func main() {
+	flag.Parse()
+	logrus.SetLevel(log.SanitizeLogLevel(*logLevel))
+	logrus.SetFormatter(&logrus.TextFormatter{FullTimestamp: true})
+	logrus.Printf("getbackup pod starting, git commit %s", gitCommit)
```
Comment (kargakis) [inline cmd/getbackup/getbackup.go]: this app shouldn't really care whether it runs in a pod or not, also you seem to run it as a bare container.

File: pkg/arm/data/master-startup.sh [hunk 10]
```diff
@@ -659,6 +659,16 @@ sed -i -re "s#( *server: ).*#\1https://$(hostname)#" /etc/origin/node/node.kubec
 # daemonset will not fail and set the master node labels correctly.
 cp /etc/origin/node/node.kubeconfig /etc/origin/node/bootstrap.kubeconfig
 
+{{- if $.Extra.IsRecovery }}
+tempBackDir=$(mktemp -d -t etcd-backup)
+docker run -v /etc/origin/cloudprovider/:/_data/_out -v $tempBackDir:/out:z {{ .Config.Images.GetBackup }} --blobname=$.Extra.BackupBlobName --destination=/out/backup.db
```
Comment (kargakis) [inline pkg/arm/data/master-startup.sh]: @aaronlevy fyi, this is a proposed solution for how we are going to restore from backups. Is this something handled during Ignition in the 4.x installer today? If not, how are you thinking of doing restores?

File: pkg/addons/data/ConfigMap/openshift-etcd/backup-config.yaml [hunk 11]
```diff
@@ -0,0 +1,24 @@
+apiVersion: v1
```
Comment (kargakis) [inline pkg/addons/data/ConfigMap/openshift-etcd/backup-config.yaml]: Do we need to keep `pkg/addons/data/EtcdBackup.etcd.database.coreos.com/openshift-etcd/etcd-backup.yaml` with this?

File: pkg/addons/data/CronJob.batch/openshift-etcd/etcd-backup-cleaner.yaml [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  name: etcd-backup-cleaner
+  namespace: openshift-etcd
+spec:
+  # todo: move to 1 day after testing
+  schedule: "0 * * * *"
+  jobTemplate:
+    spec:
+      template:
+        spec:
+          serviceAccountName: etcd-operator
```
Comment (kargakis) [inline pkg/addons/data/CronJob.batch/openshift-etcd/etcd-backup-cleaner.yaml]: The etcd-operator has elevated access where this cronjob only needs to get and delete an etcdbackup. Needs its own service account, role, and rolebinding.

File: pkg/addons/data/CronJob.batch/openshift-etcd/etcd-backup.yaml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  name: etcd-backup
+  namespace: openshift-etcd
+spec:
+  # todo: move to 1 hour after testing
+  schedule: "*/10 * * * *"
+  jobTemplate:
+    spec:
+      template:
+        spec:
+          serviceAccountName: etcd-operator
```
Comment (kargakis) [inline pkg/addons/data/CronJob.batch/openshift-etcd/etcd-backup.yaml]: The etcd-operator has elevated access where this cronjob only needs to create an etcdbackup. Needs its own service account, role, and rolebinding.

File: pkg/addons/data/ConfigMap/openshift-etcd/backup-config.yaml (hunk 11 above)
Comment (kargakis) [reply pkg/addons/data/ConfigMap/openshift-etcd/backup-config.yaml]: Nevermind, I see you remove this in this PR.

File: pkg/config/images.go [hunk 14]
```diff
@@ -97,6 +98,7 @@ func (g *simpleGenerator) selectContainerImagesOrigin(cs *api.OpenShiftManagedCl
 		c.Images.RegistryConsole = "docker.io/cockpit/kubernetes:latest"
 		c.Images.Sync = "quay.io/openshift-on-azure/sync:v3.10"
 		c.Images.LogBridge = "quay.io/openshift-on-azure/logbridge:latest"
+		c.Images.GetBackup = "quay.io/openshift-on-azure/getbackup:latest"
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/config/images.go]: all it does is download a blob to a local file

File: Dockerfile.getbackup (hunk 1 above)
Comment (asalkeld) [reply Dockerfile.getbackup]: done

File: cmd/getbackup/getbackup.go (hunk 2 above)
Comment (asalkeld) [reply cmd/getbackup/getbackup.go]: thanks

File: cmd/getbackup/getbackup.go (hunk 3 above)
Comment (asalkeld) [reply cmd/getbackup/getbackup.go]: hehe,. oops

File: pkg/fakerp/fakerp.go (hunk 5 above)
Comment (asalkeld) [reply pkg/fakerp/fakerp.go]: ok, i'll revert. My thought was to have the read/write next to each other for maintainability.

File: cmd/getbackup/getbackup.go (hunk 7 above)
Comment (asalkeld) [reply cmd/getbackup/getbackup.go]: done

File: cmd/getbackup/getbackup.go (hunk 8 above)
Comment (asalkeld) [reply cmd/getbackup/getbackup.go]: yeah, debug artifact

File: cmd/getbackup/getbackup.go (hunk 9 above)
Comment (asalkeld) [reply cmd/getbackup/getbackup.go]: ack

**Comment by asalkeld** [comment]
so after a bit of testing, I am going to have to solve the hostname problem. (when you delete the vms and update the hostnames just keep incrementing). So my plan is to make another little binary that runs in a container that lists the vms in the scale set and does the following
(in master-startup.sh before any use of $(hostname))

# call the container: "what-should-my-hostname-be" $(hostname)
# it lists the vms in the scale set (and waits until there are 3) and sorts them by name
# it then returns what this host should be called
# eg:
# master-03 -> master-00
# master-07 -> master-01
# master-23 -> master-02
we can then set the hostname - getting stable hostnames on the masters

cc @jim-minter

File: pkg/fakerp/fakerp.go (hunk 5 above)
Comment (kargakis) [reply pkg/fakerp/fakerp.go]: Agreed, they are all over the place, we may want to revive https://github.com/openshift/openshift-azure/issues/642

**Comment by kargakis** [comment]
@asalkeld can this be closed in favor of https://github.com/openshift/openshift-azure/pull/713?

**Comment by asalkeld** [comment]
> @asalkeld can this be closed in favor of #713?

sure, i'll open a new one with just the recovery part once the backup is merged
